spec/ics-2-consensus-requirements/README.md

```diff
@@ -0,0 +1,124 @@
+conflicting `Commit`s can be both validated, thus no data can be rewritten after it has been 
+committed. Two `Commit`s are conflicting when both has same height but not equal.
+
+* Verifiers can verify future `Commit`s using an existing `TrustedCommit`. When the verifier 
+validates it, the verified header is in the canonical blockchain.
+
+* `TrustedCommit`s contains an accumulator root(ICS23) that the other logics can verify whether 
+key-value pairs exists or not with it.
+
+### Technical Specification
+
+#### Definitions
+
+* `TrustedCommit` is a blockchain commit which can be used to prove future commits, stored in
+  one blockchain to verify the state of the other.
+  Defined as 3-tuple `(v :: Commit -> (Error|TrustedCommit), r :: AccumulatorRoot)`, where
+    * `v` is the verifier, proves child `Commit.p` and returns the updated `TrustedCommit`
+    * `r` is the `AccumuatorRoot`, used to prove internal state
+
+* `Commit` is a blockchain header which provides information to update `TrustedCommit`, 
+  submitted to one blockchain to update the stored `TrustedCommit`.
+  Defined as 2-tuple `(p :: CommitProof, tc :: TrustedCommit`, where
+    * `p` is the commit proof used by `TrustedCommit.v` to verify
+    * `tc` is the new `TrustedCommit` which will replace the existing one after the verification
+ 
+* `Consensus` is a blockchain consensus algorithm which generates valid `Commit`s.
+  Defined as a function `Commit -> [Transaction] -> Commit`
+
+* `Blockchain` is a subset of `(TrustedCommit, [Commit])`, generated by a `Consensus`.
+
+#### Requirements
+
+Consensus verification requires the following accumulator primitives with datastructures and
+properties as defined in ICS23:
+
+* `AccumulatorRoot`
+
+#### Subprotocols
+
+##### Verifier
+
+Verifiers prove new `Commit` generated from a blockchain. It is expected to prove 
+a `Commit` efficiently; efficient then replaying `Consensus` logic for given parent `Commit`
+and the transactions. 
+
+##### Accumulator Root
+
+`TrustedCommit` contains an accumulator root, which identifies the whole state of the corresponding
+blockchain at the point of time that the commit is generated. It is expected that the verifying 
+inclusion or exclusion of certain data in the accumulator is done efficient. See ICS23 for the details
+about the `AccumulatorRoot`s.
+
+##### CommitProof
+
+`CommitProof` is a type dependent on concrete implementations of `Commit`. `TrustedCommit.v` uses
+the information stored in `Commit.p` to check `Commit.tc`. 
+
+##### Consensus 
+
+`Consensus` is a blockchain protocol which actually generates a list of `Commit`. 
+
+1. Headers have no more than one direct child
+ 
+* Satisfied if: deterministic safety
+* Possible violation scenario: validator double signing, miner double spend
+
+2. Headers have at least one direct child
+
+* Satisfied if: liveness, lightclient verifier continuability
+* Possible violation scenario: synchronized halt, incompatible hard fork
```

Comment:
What is a synchronized halt?

Reply:
More than 1/3 of the validators goes offline at once. Can be happened on a less decentralized network.

Reply:
Is that actually a requirement of safety here?

I think we need to be more specific. Certainly all headers won't have a direct child at any point in time, as there will be one final header (the head of the chain).

Ideally we could analyze what happens under safety and liveness faults separately - the latter should be much less damaging, maybe some problems with timeouts.

Reply:
Per discussion: the packet semantics(not connection semantics) are relying on the partial synchrony assumption on the packets since the timeout logic is depending on the receiving chain's height. We should include that in the protocol. The damage analysis and the recovery methods on the safety/liveness fault should be separated, the amount of impact and the method that has to be applied in each case are pretty different. The safety can be addressed in #6 and #53.

Reply:
Opened #63 for liveness